plasma_framework/contracts/src/fast_exits/liquidity.sol

```diff
@@ -0,0 +1,202 @@
+            "First Tx Provided was not called by sender"
+        );
+    }
+
+    /**
+     * @notice Verify the First Tx provided is valid
+     * @param _utxoPosInput position of the output that created the inputs for second transaction
+     * @param _rlpInputCreationTx RLP-encoded first transaction that transfers to this contract
+     * @param _inputCreationTxInclusionProof First transactions inclusion proofs
+    */
+    function verifyTxValidity(
+        uint256 _utxoPosInput,
+        bytes memory _rlpInputCreationTx,
+        bytes memory _inputCreationTxInclusionProof
+    ) internal {
+        PosLib.Position memory utxoDecoded = PosLib.decode(_utxoPosInput);
+        utxoDecoded.outputIndex = 0;
+        (bytes32 _root, ) = pf.blocks(utxoDecoded.blockNum);
+        require(_root != bytes32(""), "Failed to get root of the block");
+        bool txExists = Merkle.checkMembership(
+            _rlpInputCreationTx,
+            utxoDecoded.txIndex,
+            _root,
+            _inputCreationTxInclusionProof
+        );
+        require(
+            txExists,
+            "Provided Transaction isn't finalized or doesn't exist"
+        );
+    }
+
+    /**
+     * @notice func that calls omg-contracts to start the exit
+     * @param _utxoPosToExit position of the output which the contract has to exit
+     * @param _rlpOutputTxToContract RLP-encoded transaction that creates the output for the contract
+     * @param _outputTxToContractInclusionProof Second Transaction's inclusion proof
+    */
+    function runExit(
+        uint256 _utxoPosToExit,
+        bytes memory _rlpOutputTxToContract,
+        bytes memory _outputTxToContractInclusionProof
+    ) internal {
+        PaymentStandardExitRouterArgs.StartStandardExitArgs memory s;
+        s.utxoPos = _utxoPosToExit;
+        s.rlpOutputTx = _rlpOutputTxToContract;
+        s.outputTxInclusionProof = _outputTxToContractInclusionProof;
+        peg.startStandardExit.value(msg.value)(s);
+    }
+
+    function getExitId(bytes memory _txBytes, uint16 _outputIndex)
+        internal
+        pure
+        returns (uint160)
+    {
+        uint256 exitId = (uint256(keccak256(_txBytes)) >> 105) |
+            (uint256(_outputIndex) << 152);
+        uint160 croppedExitId = uint160(exitId);
+        require(uint256(croppedExitId) == exitId, "ExitId overflows");
+        return croppedExitId;
+    }
+
+    /**
+     * @dev Get Amount from contract after exit is processed - (to be updated)
+     * @param _exitId The exit id
+    */
+    function getWithdrawl(uint160 _exitId) public {
+        require(
+            exitIdtoUser[_exitId] == msg.sender,
+            "Only the exitId owner can get the withdrawl"
+        );
+        uint160[] memory _exitIdList = new uint160[](1);
+        _exitIdList[0] = _exitId;
+        PaymentExitDataModel.StandardExit[] memory exits = peg.standardExits(
+            _exitIdList
+        );
+        if (exits[0].utxoPos == 0) {
```

Comment:
this is quite interesting solution of verifying the exit has been processed. 
However, I *think* that if the exit got challenge, it would be the same like this as it will be deleted.

Worth to see if you can find some good solution to recognize an exit is finalized, if you find out that making change to existing contracts is required or would be better solution, now (this cycle) is the time to shout it out for discussion!

Reply:
@boolafish this is a very good point and sth that is actually a blocker - otherwise users' funds may be easily stolen.

Reply:
Let's move the discussion to #628